File: dis-25/z.ml
type 'a recurse =
| Recurse of ('a recurse -> 'a -> 'a)

let pre_fact (Recurse f: int recurse) (n: int) =
  if n = 0 then 1 else n * ((f (Recurse f)) (n - 1))

let fact =
  pre_fact (Recurse pre_fact)

File: dis-25/z.py
def z(f):
    return (lambda x: f(lambda y: x(x)(y)))(lambda x: f(lambda y: x(x)(y)))


def pre_fact(f):
    return lambda n: 1 if n == 0 else n * f(n - 1)


def fact():
    return z(pre_fact)


print(fact()(3))
